File: src/core/feriados.py
"""
Feriados legales de Chile para pintarlos en el calendario de la agenda.

Fuente: https://feriados-cl.netlify.app/api/holidays/<año> (solo responde
para el año en curso en adelante: un año pasado devuelve 400).

Tres capas, de más a menos fresca, y ninguna puede voltear la app -- un
calendario sin feriados marcados es un detalle visual, no un error:

  1. resources/local_cache/feriados_<año>.json -- lo que ya se bajó alguna
     vez. Si existe se usa tal cual y NO se vuelve a pedir por red: la lista
     de un año no cambia (salvo ley nueva, ver refrescar()).
  2. la API, en un hilo aparte (ver FeriadosThread) para no congelar la
     ventana si netlify no responde.
  3. resources/json/feriados_backup.json -- copia versionada en el repo,
     para una instalación nueva sin red.

El formato normalizado es {"MM-DD": "descripción"} por año; el backup es
{"<año>": {...}} con esos mismos diccionarios adentro.
"""
import json
import os
from pathlib import Path

import requests
from PySide6.QtCore import QThread, Signal

FERIADOS_URL = "https://feriados-cl.netlify.app/api/holidays/{year}"
FERIADOS_TIMEOUT = 5

_RESOURCES = Path(__file__).resolve().parents[2] / 'resources'
# Misma convención que core/app_layout.py: local_cache/ es data dinámica del
# usuario y el updater la preserva entre versiones.
CACHE_DIR = _RESOURCES / 'local_cache'
BACKUP_PATH = _RESOURCES / 'json' / 'feriados_backup.json'


def _cache_path(year: int) -> Path:
    return CACHE_DIR / f'feriados_{year}.json'


def _es_mapa_feriados(data) -> bool:
    return (isinstance(data, dict)
            and all(isinstance(k, str) and isinstance(v, str) for k, v in data.items()))


def _parse_respuesta(data) -> dict:
    """{"feriados": {"enero": [{"mes": 1, "dia": 1, "descripcion": ...}]}}
    -> {"01-01": "Año Nuevo"}. Devuelve {} si el JSON no tiene esa forma."""
    feriados = data.get("feriados") if isinstance(data, dict) else None
    if not isinstance(feriados, dict):
        return {}

    mapa = {}
    for dias in feriados.values():
        if not isinstance(dias, list):
            continue
        for dia in dias:
            if not isinstance(dia, dict):
                continue
            try:
                mes, num = int(dia["mes"]), int(dia["dia"])
            except (KeyError, TypeError, ValueError):
                continue
            if not (1 <= mes <= 12 and 1 <= num <= 31):
                continue
            mapa[f"{mes:02d}-{num:02d}"] = str(dia.get("descripcion") or "Feriado")
    return mapa


def _leer_json(path: Path):
    try:
        with open(path, 'r', encoding='utf-8') as fh:
            return json.load(fh)
    except (OSError, ValueError):
        return None


def load_cache(year: int) -> dict | None:
    """Feriados del año ya guardados en disco, o None si no hay/está roto."""
    data = _leer_json(_cache_path(year))
    return data if _es_mapa_feriados(data) else None


def save_cache(year: int, mapa: dict) -> None:
    """Escritura atómica (tmp + replace): un json a medio escribir sería
    justo el que se lee en el próximo arranque sin red."""
    try:
        CACHE_DIR.mkdir(parents=True, exist_ok=True)
        tmp = _cache_path(year).with_suffix('.json.tmp')
        with open(tmp, 'w', encoding='utf-8') as fh:
            json.dump(mapa, fh, ensure_ascii=False, indent=2, sort_keys=True)
        os.replace(tmp, _cache_path(year))
    except OSError:
        pass  # disco lleno / permisos: no es motivo para romper la agenda


def load_backup(year: int) -> dict:
    """Copia versionada en el repo. {} si no trae ese año."""
    data = _leer_json(BACKUP_PATH)
    if not isinstance(data, dict):
        return {}
    mapa = data.get(str(year))
    return mapa if _es_mapa_feriados(mapa) else {}


def fetch_from_network(year: int) -> dict:
    """Un intento contra la API. {} ante cualquier falla (red, 400 de un año
    pasado, JSON raro) -- nunca lanza. Si trae algo, refresca la cache."""
    try:
        resp = requests.get(FERIADOS_URL.format(year=year), timeout=FERIADOS_TIMEOUT)
    except requests.RequestException:
        return {}
    if resp.status_code >= 400:
        return {}
    try:
        mapa = _parse_respuesta(resp.json())
    except ValueError:
        return {}
    if mapa:
        save_cache(year, mapa)
    return mapa


def feriados_offline(year: int) -> dict:
    """Lo que se puede saber sin tocar la red: cache, o el backup del repo."""
    return load_cache(year) or load_backup(year)


def refrescar(year: int) -> dict:
    """Feriados del año pidiéndolos por red solo si no hay cache. Bloquea:
    llamar desde FeriadosThread, no desde el hilo de UI.

    Sin TTL a propósito: los feriados de un año ya bajado no cambian. Si
    algún año se agrega uno por ley, basta borrar resources/local_cache/
    feriados_<año>.json (o actualizar el backup del repo)."""
    cache = load_cache(year)
    if cache:
        return cache
    return fetch_from_network(year) or load_backup(year)


class FeriadosThread(QThread):
    """Trae los feriados de varios años fuera del hilo de UI (netlify caído
    congelaría la agenda hasta FERIADOS_TIMEOUT por año). Emite una sola vez
    con {año: {"MM-DD": descripción}}; si no consiguió nada, no emite."""

    listo = Signal(dict)

    def __init__(self, years, parent=None):
        super().__init__(parent)
        self._years = list(years)

    def run(self) -> None:
        resultado = {}
        for year in self._years:
            if self.isInterruptionRequested():
                return
            mapa = refrescar(year)
            if mapa:
                resultado[year] = mapa
        if resultado:
            self.listo.emit(resultado)

File: labsim_backend/src/Feriados.php
<?php

declare(strict_types=1);

/**
 * Feriados legales de Chile para marcarlos en el calendario de agenda.php --
 * agendar una cita en feriado es un error del docente que hoy no avisa nada.
 *
 * Fuente: https://feriados-cl.netlify.app/api/holidays/<año> (responde 400
 * para años pasados). Mismas tres capas que el cliente de escritorio
 * (src/core/feriados.py), y ninguna puede voltear la página:
 *
 *   1. data/feriados_<año>.json -- si existe se usa y NO se pide nada por red
 *      (la lista de un año ya bajado no cambia).
 *   2. la API, con timeout corto.
 *   3. resources/feriados_backup.json -- copia versionada en el repo, para un
 *      hosting sin salida a internet.
 *
 * Formato normalizado: ['MM-DD' => 'descripción'] por año.
 */
class Feriados
{
    private const URL = 'https://feriados-cl.netlify.app/api/holidays/%d';
    private const TIMEOUT = 4;

    /** @var array<int, array<string, string>> memo por request */
    private static array $memo = [];

    /**
     * Feriados del año. [] si no se pudo conseguir por ningún lado -- el
     * calendario se dibuja igual, solo que sin marcas.
     *
     * @return array<string, string>
     */
    public static function delAnio(int $year): array
    {
        if (isset(self::$memo[$year])) {
            return self::$memo[$year];
        }

        $mapa = self::leerJson(self::cachePath($year));
        if ($mapa === null) {
            $mapa = self::fetch($year);
            if ($mapa !== []) {
                self::guardarCache($year, $mapa);
            }
        }
        if ($mapa === null || $mapa === []) {
            $mapa = self::backup($year);
        }

        self::$memo[$year] = $mapa;
        return $mapa;
    }

    /**
     * Feriados de varios años en un solo arreglo ['YYYY-MM-DD' => descripción],
     * que es la forma en que los usa el calendario (y el JS del modal).
     *
     * @param int[] $years
     * @return array<string, string>
     */
    public static function porFecha(array $years): array
    {
        $porFecha = [];
        foreach (array_unique($years) as $year) {
            foreach (self::delAnio((int) $year) as $mmdd => $descripcion) {
                $porFecha[sprintf('%04d-%s', $year, $mmdd)] = $descripcion;
            }
        }
        return $porFecha;
    }

    private static function cachePath(int $year): string
    {
        return __DIR__ . '/../data/feriados_' . $year . '.json';
    }

    /** @return array<string, string>|null */
    private static function leerJson(string $path): ?array
    {
        if (!is_file($path)) {
            return null;
        }
        $raw = @file_get_contents($path);
        if ($raw === false) {
            return null;
        }
        $data = json_decode($raw, true);
        return self::esMapa($data) ? $data : null;
    }

    private static function esMapa($data): bool
    {
        if (!is_array($data)) {
            return false;
        }
        foreach ($data as $k => $v) {
            if (!is_string($k) || !is_string($v)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Un intento contra la API. [] ante cualquier falla (sin red, 400 de un
     * año pasado, JSON raro) -- nunca lanza.
     *
     * @return array<string, string>
     */
    private static function fetch(int $year): array
    {
        if (!function_exists('curl_init')) {
            return [];
        }
        $ch = curl_init(sprintf(self::URL, $year));
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => self::TIMEOUT,
            CURLOPT_CONNECTTIMEOUT => self::TIMEOUT,
        ]);
        $body = curl_exec($ch);
        $status = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($body === false || $status < 200 || $status >= 300) {
            return [];
        }
        return self::parse(json_decode((string) $body, true));
    }

    /**
     * {"feriados": {"enero": [{"mes": 1, "dia": 1, "descripcion": ...}]}}
     * -> ['01-01' => 'Año Nuevo'].
     *
     * @return array<string, string>
     */
    public static function parse($data): array
    {
        $feriados = is_array($data) ? ($data['feriados'] ?? null) : null;
        if (!is_array($feriados)) {
            return [];
        }

        $mapa = [];
        foreach ($feriados as $dias) {
            if (!is_array($dias)) {
                continue;
            }
            foreach ($dias as $dia) {
                if (!is_array($dia) || !isset($dia['mes'], $dia['dia'])) {
                    continue;
                }
                $mes = (int) $dia['mes'];
                $num = (int) $dia['dia'];
                if ($mes < 1 || $mes > 12 || $num < 1 || $num > 31) {
                    continue;
                }
                $descripcion = isset($dia['descripcion']) ? trim((string) $dia['descripcion']) : '';
                $mapa[sprintf('%02d-%02d', $mes, $num)] = $descripcion !== '' ? $descripcion : 'Feriado';
            }
        }
        return $mapa;
    }

    /** @param array<string, string> $mapa */
    private static function guardarCache(int $year, array $mapa): void
    {
        $path = self::cachePath($year);
        $dir = dirname($path);
        if (!is_dir($dir) && !@mkdir($dir, 0775, true) && !is_dir($dir)) {
            return;
        }
        // tmp + rename: un json a medio escribir sería justo el que se lee en
        // el próximo request.
        $tmp = $path . '.tmp';
        $json = json_encode($mapa, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
        if ($json === false || @file_put_contents($tmp, $json) === false) {
            return;
        }
        @rename($tmp, $path);
    }

    /** @return array<string, string> */
    private static function backup(int $year): array
    {
        $data = json_decode((string) @file_get_contents(__DIR__ . '/../resources/feriados_backup.json'), true);
        if (!is_array($data) || !isset($data[(string) $year])) {
            return [];
        }
        $mapa = $data[(string) $year];
        return self::esMapa($mapa) ? $mapa : [];
    }
}
